JDK-8293156: Dcmd VM.classloaders fails to print the full hierarchy

tstuefe opened the pull request:
Fixes a bug in the `VM.classloaders` jcmd that causes class loaders to be omitted from the output if a parent class loader never loaded any class and therefore had no associated DCmd.

The fix changes the command to not rely on the existence of a CLD structure for the loader; instead, all information (loader class name, loader name, etc) is pulled via the loader oop, which has to be always there unless its the bootstrap loader.

Also, the tests were expanded to test the display of empty loaders and empty parent loaders.

Thanks to @dholmes-ora for finding this bug.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293156](https://bugs.openjdk.org/browse/JDK-8293156): Dcmd VM.classloaders fails to print the full hierarchy


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10312/head:pull/10312` \
`$ git checkout pull/10312`

Update a local copy of the PR: \
`$ git checkout pull/10312` \
`$ git pull https://git.openjdk.org/jdk pull/10312/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10312`

View PR using the GUI difftool: \
`$ git pr show -t 10312`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10312.diff">https://git.openjdk.org/jdk/pull/10312.diff</a>

</details>

plummercj commented on src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:
```diff
@@ -154,33 +154,59 @@ class LoaderTreeNode : public ResourceObj {
+    return false;
+  }
+
   void print_with_childs(outputStream* st, BranchTracker& branchtracker,
```
I know this isn't part of your changes, but the plural of child is children, not childs. However, I think child_nodes would probably be best here.

tstuefe replied:
Fixed

plummercj replied:
It looks like there are a couple more references to "childs" still present.

plummercj commented on src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:
```diff
@@ -287,8 +315,9 @@ class LoaderTreeNode : public ResourceObj {
     assert(is_leaf() && target_node->is_leaf(), "must be leaf");
-    return _cld->class_loader_klass() == target_node->_cld->class_loader_klass() &&
-           _cld->name() == target_node->_cld->name();
+    return (loader_klass() != NULL) && (loader_klass() == target_node->loader_klass()) &&
+           ((loader_name_oop() == NULL && target_node->loader_name_oop() == NULL) ||
+             (::strcmp(loader_name(), target_node->loader_name()) == 0));
```
This is hard to parse. I think using if/else statements would be cleaner.

tstuefe replied:
Fixed

plummercj commented on test/hotspot/jtreg/serviceability/dcmd/vm/ClassLoaderHierarchyTest.java:
```diff
@@ -60,39 +73,77 @@ public class ClassLoaderHierarchyTest {
-        output.shouldMatch("Kevin.*TestClassLoader");
-        output.shouldMatch("Bill.*TestClassLoader");
+        // (A)
+        output.shouldContain("+-- <bootstrap>");
+        output.shouldContain("      +-- \"platform\", jdk.internal.loader.ClassLoaders$PlatformClassLoader");
+        output.shouldContain("      |     +-- \"app\", jdk.internal.loader.ClassLoaders$AppClassLoader");
+        output.shouldContain("      +-- \"Kevin\", ClassLoaderHierarchyTest$TestClassLoader");
+        output.shouldContain("      +-- ClassLoaderHierarchyTest$TestClassLoader");
+        output.shouldContain("      |     +-- \"Bill\", ClassLoaderHierarchyTest$TestClassLoader");
```
Can you test using one long string, and do the same for each section below? I think embedding \\R as the newline character should work on all platforms. If you run into unforeseen issues, then don't worry about.

tstuefe replied:
Fixed

tstuefe commented:
Hi @plummercj , thanks for your Review. I worked most of your your feedback.

.. Thomas

dholmes-ora commented on src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp:
```diff
@@ -154,33 +154,59 @@ class LoaderTreeNode : public ResourceObj {
+        st->print(" \"%s\",", the_loader_name);
       }
-      st->print(" %s", loader_klass != NULL ? loader_klass->external_name() : "??");
+      st->print(" %s", the_loader_class_name);
```
`the_loader_class_name` could be NULL here. You need to change `loader_class_name()` t return "??" like the original code uses.

tstuefe replied:
Good catch. We never encounter this because Klass can only be null for the bootstrap loader, which is handled in a different branch. I'll fix it.

dholmes-ora reviewed (changes requested):
Hi Thomas,

Generally this seems fine - but one issue below.

Thanks.

tstuefe commented:
> Hi @plummercj , thanks for your Review. I worked most of your your feedback.
> 
> .. Thomas

tstuefe commented:
- Fixed all remaining occurrences of "childs"
- Fixed possible segfault when printing class name for bootstrap loader
- Fixed GHAs (apparently we don't have a MainWrapper class when executing jtreg in GHAs, so I cannot parse for it)

tstuefe commented:
Thanks @plummercj and @dholmes-ora .

/integrate
